bootleg/utils/classes/dictvocabulary_tries.py

```diff
@@ -0,0 +1,361 @@
+"""Alias to candidate trie."""
+import itertools
+import json
+import os
+from typing import Any, Callable, Dict, List, Set, Tuple, Union
+
+import marisa_trie
+import ujson
+from tqdm import tqdm
+
+
+def flatten(arr):
+    """Flatten array."""
+    return [item for sublist in arr for item in sublist]
+
+
+def load_json_file(filename):
+    """Load json file."""
+    with open(filename, "r") as f:
+        contents = ujson.load(f)
+    return contents
+
+
+def dump_json_file(filename, contents):
+    """Dump json file."""
+    with open(filename, "w") as f:
+        try:
+            ujson.dump(contents, f)
+        except OverflowError:
+            json.dump(contents, f)
+
+
+def get_cand_with_score(
+    max_value: int, value: List[Tuple[str, int]], vocabulary: marisa_trie
+):
+    """Get the keys with score numerical values as list of ints."""
+    assert type(value) is list
+    if len(value) > 0:
+        assert all(type(v[0]) is str for v in value)
+        assert all((type(v[1]) is float) or (type(v[1]) is int) for v in value)
+    new_value = flatten([[vocabulary[p[0]], p[1]] for p in value])[: (2 * max_value)]
+    assert -1 not in new_value
+    overflowed = len(new_value) == (2 * max_value)
+    new_value.extend([-1] * (2 * max_value - len(new_value)))
+    return tuple(new_value), overflowed
+
+
+def get_key_value_pair(
+    max_value: int,
+    value: List[Tuple[str, int]],
+    key_vocabulary: marisa_trie,
+    value_vocabulary: marisa_trie,
+):
+    """Get the key value pairs as list of ints."""
+    new_value = flatten(
+        [[key_vocabulary[p[0]], value_vocabulary[p[1]]] for p in value]
+    )[: (2 * max_value)]
+    assert -1 not in new_value
+    overflowed = len(new_value) == (2 * max_value)
+    new_value.extend([-1] * (2 * max_value - len(new_value)))
+    return tuple(new_value), overflowed
+
+
+def inverse_qid_cand_with_score(value: List[int], itos: Callable[[int], str]):
+    """Return entity candidate and prob score from numerical values."""
+    assert len(value) % 2 == 0
+    new_value = []
+    for i in range(0, len(value), 2):
+        # -1 values are only added at the end as padding
+        if value[i] == -1:
+            break
+        new_value.append([itos(value[i]), value[i + 1]])
+    return new_value
+
+
+def inverse_key_value_pair(
+    value: List[int], key_itos: Callable[[int], str], value_itos: Callable[[int], str]
+):
+    """Return list of key value pairs from numerical values."""
+    assert len(value) % 2 == 0
+    new_value = []
+    for i in range(0, len(value), 2):
+        # -1 values are only added at the end as padding
+        if value[i] == -1:
+            break
+        new_value.append([key_itos(value[i]), value_itos(value[i + 1])])
+    return new_value
+
+
+class TwoLayerVocabularyScoreTrie:
+    """TwoLayerVocabularyScoreTrie.
+
+    This creates a record tri from a string to a list of string candidates. These candidates are either a single
```

Comment:
It's a Trie data structure. A standard tree structure for string keys. Or are you saying I should rename it to be something else because trie is confusing?